### ISSUE TITLE
Fresh install of Doom failed on installing of fsharp (worked on previous installation)

### ISSUE THREAD

**Issue by geoffder** [hlissner/doom-emacs#3650]
**What did you expect to happen?**
For doom install (and subsequent attempts to sync) to install fsharp-mode.

**What actually happened?**
```
- Debug mode on
> Synchronizing your config with Doom Emacs...
  > Regenerating envvars file at "~/.emacs.d/.local/env"
    - Ignoring DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
    - Ignoring TERM=xterm-kitty
    - Ignoring HOME=/home/geoff
    - Ignoring PWD=/home/geoff
    ✓ Successfully generated "~/.emacs.d/.local/env"

DOOM Initializing package.el
DOOM Initializing straight.el
  > Installing packages...
    → Building fsharp-mode...
    → Building fsharp-mode → Building eglot...
    → Building fsharp-mode → Building eglot → Building jsonrpc...
    → Building fsharp-mode → Building eglot → Building jsonrpc...done
    → Building fsharp-mode → Building eglot → Building flymake...
    → Building fsharp-mode → Building eglot → Building flymake → Building eldoc...
    → Building fsharp-mode → Building eglot → Building flymake → Building eldoc...done
    → Building fsharp-mode → Building eglot → Building flymake...
    → Building fsharp-mode → Building eglot → Building flymake...done
    → Building fsharp-mode → Building eglot → Building project...
    → Building fsharp-mode → Building eglot → Building project...done
x There was an unexpected error:
  Type: doom-package-error
  Message:
    Error with packages
  Data:
    "fsharp-mode"
    (void-variable tab-prefix-map)
```

**Additional details:**
fsharp lines are commented out of init.el right now due to the present issue.
https://github.com/geoffder/configs/tree/master/confs/home/geoff/.doom.d

<details><pre>
If an error message is involved include a backtrace of it.

How to acquire a backtrace:
https://github.com/hlissner/doom-emacs/blob/develop/docs/getting_started.org#how-to-extract-a-backtrace-from-an-error
</pre></details>

**Steps to reproduce:**
1. Uncomment fsharp in init.el
2. doom resync

**System information:**
<details><pre>
```
emacs   version    26.3
        features   XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GSETTINGS GLIB NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD LCMS2
        build      May 19, 2020
        buildopts  (--prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --with-x-toolkit=gtk3 --with-xft --with-wide-int --with-modules 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now)
        windowsys  x
        daemonp    server-running
doom    version    2.0.9
        build      grafted, HEAD -> develop, origin/develop, origin/HEAD 96ab60d 2020-07-29 16:10:31 -0400
        dir        ~/.doom.d/
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /bin/bash
        uname      Linux 5.4.52-1-MANJARO #1 SMP PREEMPT Thu Jul 16 16:07:11 UTC 2020 x86_64
        path       (~/.opam/default/bin ~/miniconda3/bin ~/miniconda3/condabin ~/.local/bin /usr/local/bin /usr/bin /bin /usr/local/sbin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /var/lib/snapd/snap/bin ~/neuron/nrn/x86_64/bin /usr/lib/emacs/26.3/x86_64-pc-linux-gnu/)
config  envfile    envvar-file
        elc-files  0
        modules    (:completion (company +auto) ivy :ui doom doom-dashboard fill-column hl-todo indent-guides modeline nav-flash ophints (popup +all +defaults) tabs treemacs unicode vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold (format +onsave) multiple-cursors rotate-text snippets :emacs dired electric ibuffer vc :checkers syntax :tools debugger (eval +overlay) (lookup +docsets) lsp magit :lang cc csharp data elixir emacs-lisp ess (haskell +dante) javascript julia markdown ocaml (org +dragndrop +present) (python +lsp +conda) sh :config (default +bindings +smartparens))
        packages   (n/a)
        unpin      (n/a)
        elpa       (n/a)
```
</pre></details>

**Comment by geoffder** [comment]
> How did you acquire this version of Emacs? Did you build it from source?

I got it from the Arch repository, I never noticed the recommendation to used emacs27 before, I'll give upgrading a go later.

Thanks for the quick reply!

**Comment by geoffder** [comment]
Packages are installing fine now and running error free, except now my fsharp buffers are running painfully slowly (elisp/ocaml/python that I've tested are fine). Even disabling the major mode, disconnecting the lsp / flycheck / font-lock etc don't seem to let it spring in to action either. 

Anyway, the original issue no longer exists, but I am at a loss as to how to approach this new problem :thinking: ...

**Comment by geoffder** [comment]
Disabling doom-modeline-mode seems to restore a lot of the responsiveness (no longer a 500ms+ to move cursor) in fsharp buffers. Again, no problem in other buffers with nearly all the same modes activated (including lsp in the case of python).

**Comment by hlissner** [comment]
`doom-modeline` can be unexpectedly slow. I'd recommend switching to `:ui (modeline +light)`. It will one day replace `doom-modeline`.

**Comment by hlissner** [comment]
Could you try adding this to your `~/.doom.d/packages.el`:

```elisp
(package! project :pin "76ccdcf2ea6b349a67c5aee166c16a7dc162f2e0")
```

Then run `doom sync -u`, and see if that changes anything?

**Comment by geoffder** [comment]
> `doom-modeline` can be unexpectedly slow. I'd recommend switching to `:ui (modeline +light)`. It will one day replace `doom-modeline`.

Thanks I'll try this out.

> Could you try adding this to your `~/.doom.d/packages.el`:
> 
> ```emacs-lisp
> (package! project :pin "76ccdcf2ea6b349a67c5aee166c16a7dc162f2e0")
> ```
> 
> Then run `doom sync -u`, and see if that changes anything?

I gave it a shot and it didn't seem to change anything. May I ask what this did?

I also have an update on the modeline - fsharp interaction. I narrowed it down to 
column-number-mode and line-number-mode slowing down navigation laterally and vertically respectively in normal mode. Deactivating them like so,
```emacs-lisp
(add-hook 'fsharp-mode-hook '(lambda () (column-number-mode -1) (line-number-mode -1)))
```
seems to get rid of my cursor moving slowdowns. I came across an issue (https://github.com/seagle0128/doom-modeline/issues/353) about the fsharp vs doom-modeline that I hadn't found last night and chimed in there with my findings.

I still have an issue with the fsac server (using either eglot or lsp fronts) making insert mode super slow which I have never had before building emacs27-git from the AUR, but it is doom-modeline unrelated.

**Comment by hlissner** [comment]
> I gave it a shot and it didn't seem to change anything. May I ask what changes to project might be related that pinning to an old one might be helpful?

I suspected that project.el was prematurely using the `tab-prefix-map` variable (which isn't defined in your version of Emacs), and had hoped updating it would introduce its own fix, but it didn't.

---

In any case, this is a build issue. i.e. your OS package manager is serving up a particular build of Emacs 26/27 where `tab-prefix-map` hasn't been introduced yet, but is being used anyway. Either internally or by some plugin. It's been a nightmare to debug this, because the issue appears to originate from different places in each bug report I've received about this.

So, I introduced a clumsy, but simple fix in 0924801 that should resolve your initial error. It's good enough while we wait for it to blow over upstream. Let me know if the issue persists and I'll reopen this.

**Comment by geoffder** [comment]
Thanks for the help and explanation!  

FYI I did try out your +light modeline and unfortunately the laggy navigation issue persisted while column-number-mode and line-number-mode were active, so presumably the weird fsharp interaction is with those dependencies.
